### ISSUE TITLE
Add and expose metrics

### ISSUE THREAD

**Issue by ncdc** [heptio/ark#84]
Add various metrics using Prometheus.

**Comment by timoreimann** [comment]
As announced on Slack, we (i.e., me and/or some of my team colleagues) would be keen on adding some initial support for Prometheus-based metrics collection.

Is there anything in particular that should be included from the project maintainer's / Ark community's view? The baseline for us would probably be somewhere along the lines of collecting info about

- when a backup / restore took place
- how many objects were included
- how many errors occurred

That would allow us to set up alerts like "notify me when the last successful backup is older than x time units" and "notify me when less than x objects were considered".

Implementation-wise, we usually just include the official Go Prometheus client library. So far, it has served us well.

Happy to discuss design considerations and implementation details up front.

**Comment by ncdc** [comment]
@timoreimann I think all of those are good ideas. I'm also interested in things like
- time to list all pods, services, etc
- time to write the backup tarball
- time to upload a file to object storage
- size of tarball, log file, etc
- time to run custom backup actions and restorers

+1 to the official client lib. We're using `dep` now so you can run `hack/dep-save.sh`, where any arguments you pass are handed directly to `dep ensure`.

I think it would be a good idea to see a rough sketch of how you envision integrating metrics gathering into the code base.

Thanks!

**Comment by timoreimann** [comment]
Appreciate the prompt feedback, @ncdc. I'll try to dedicate some time next week to dig into the relevant code sections and come up with a metrics master plan to post here. :-)

**Comment by ncdc** [comment]
@timoreimann this makes sense, thanks!

FYI I did stumble across this branch where someone was already doing a POC: https://github.com/heptio/ark/compare/master...chadcatlett:stab_at_metrics. Probably worth getting @chadcatlett involved in the discussion 😄 

If we do our own registry, would it be a singleton, or would we ever want or need to have multiple?

Another thing to consider is that we're using gRPC based plugins. We'll need to figure out how to get any metrics generated by plugin instances - either by somehow getting their data into the ark server process, or by sending them to a metrics gateway or to a prometheus server directly. Something to think about...

**Comment by timoreimann** [comment]
@ncdc I don't think we'll need more one than registry, so it should be a singleton basically. The design options I outlined above should only affect how (easily) we can test things IMHO.

gRPC-based plugins are an interesting point. I'm not too familiar with the plugins architecture in Ark. If the binaries are long-running just like the Ark server, then I'd think that having the server component pull in plugin metrics via a dedicated gRPC function on each Prometheus scrape interval could be an option. OTOH, if the binaries are invoked at certain times only and terminate again while the Ark server keeps on running, letting them submit their metrics through the Prometheus pushgateway makes more sense.

AFAIK, the pushgateway doesn't come with a retention period that cleans up metrics automatically. So when in doubt, the regular pull mechanism seems preferable.

@chadcatlett: I see you have already produced some code. Is there already a PR somewhere? Happy to collaborate / coordinate to get the code into Ark. :-)

**Comment by ncdc** [comment]
Currently, some plugins are long-lived, while others are very short. Our plan, however, is to make all plugins short-lived, as it should make it easier for us to handle how we manage plugins.

My ideal setup would be to have the Ark server pull metrics from a plugin just before it terminates the plugin. I'm not sure if there's an easy way to do this with Prometheus. I stumbled across

https://github.com/prometheus/client_golang/blob/e11c6ff8170beca9d5fd8b938e71165eeec53ac6/prometheus/examples_test.go#L648-L705

which uses a custom `Gatherer`, which is a model we could perhaps consider, or something like it.

I'd prefer to avoid requiring setting up an external component such as the pushgateway as the sole means of gathering plugin metrics. It just doesn't feel right to me.

**Comment by timoreimann** [comment]
Pulling metrics just before the plugin terminates sounds like a viable option if we are able to control the plugin lifecycle.

**Comment by ncdc** [comment]
Yes, we're 100% in control (assuming the plugin doesn't crash).

**Comment by jrnt30** [comment]
We just came back to this today too and would be interested in it's progress and potentially some hands on keyboard implementation in my ever abundant free time :) 

For what it's worth, I would trend towards simplicity first on this to determine what/how we want to use it much to a few of your observations @timoreimann.  Would it be viable to put in the Global registry without plugin consideration to start?  I think this would give us the biggest bang for our buck. To your point, I could see a more generalized `TracingContext` argument being valuable and perhaps expand beyond simply a Prometheus exporter there.

I haven't researched closely enough the plugin model, but I could see a somewhat simple first pass at those metrics simply being plugin level [RED](https://www.weave.works/blog/the-red-method-key-metrics-for-microservices-architecture/) metrics to start.  This would give us visibility into the duration of the Plugin execution broken out by the plugin name to start.  If it was determined we need more granular trace like information for the internals of the plugin then we could create a "proper interface" for the plugins that could allow the exposing of that data.  

I have some similar concerns about the PushGateway integration for this due to the complexity of integration and implementation that may involve. 

Would be happy to chat about this with anyone who is passionate about seeing it happen!

**Comment by ncdc** [comment]
@jrnt30 as long as we have a high-level vision of how we plan to integrate plugin process metrics into the overall metrics registry, I would be fine starting with just the Ark server. But I would hesitate to use the global registry if we can identify up front that it's impossible to integrate metrics from external sources into it easily.

**Comment by SleepyBrett** [comment]
As a user I'd love a topline metric for each schedule that just counts backup successes and failures. As I just discovered our backups have been failing for weeks by stumbling upon it.

**Comment by rosskukulinski** [comment]
I'm not familiar enough with the Ark plugin model nor the prometheus client to weigh in on those topics.

That said, I think it's important to differentiate between the consumers of Ark metrics.  

@ncdc - as an Ark developer, you likely care about visibility and performance metrics of Ark internals based on the environment it's running in (e.g. how many objects to backup, and how long does it take)

A plugin developer likely will want to have their own metrics as well that are focused on their development process.  These metrics need to be namespaced so that two plugins don't collide on metrics.  (namespacing could be via different metric name, or via Labeling)

Ark users, however, probably care more about RED metrics (Rates - how many backups have been taken, Errors - how many backups failed, Distribution - how long did it take for backups to complete).  @SleepyBrett also makes an excellent point that we may want to have metrics labeled by Backup name and/or Scheduled name.

Quick pass at metrics & types, feedback welcome

```
# Metric labels for all include:
# * {backup="foo"} -- The name of the backup
# * {schedule="bar"} -- The name of the schedule that triggered this backup
# ark_backup totals for overall success/failure tracking
ark_backup_attempts - counter
ark_backup_completions - counter
ark_backup_failures - counter

# ark_backup_seconds describes the duration of a backup from start to finish
ark_backup_seconds -- histogram
ark_backup_seconds_count -- counter
ark_backup_seconds_sum -- counter

# ark_backup_size describes the size of the resulting tarball
ark_backup_tarball_size_bytes - histogram
ark_backup_tarball_size_count - counter
ark_backup_tarball_size_sum - counter

# ark_backup_tarball_write_seconds describes the time to write the tarball to disk
ark_backup_tarball_write_seconds
ark_backup_tarball_write_seconds_count
ark_backup_tarball_write_seconds_sum

# ark_backup_tarball_upload_seconds describes the time to upload the tarball
ark_backup_tarball_upload_seconds
ark_backup_tarball_upload_seconds_count
ark_backup_tarball_upload_seconds_sum

# ark_backup_actions_seconds describes the time to run custom backup actions
# Extra labels: action_name
ark_backup_actions_seconds -- histograms
ark_backup_actions_seconds_count -- counter
ark_backup_actions_seconds_sum -- counter

### Now restores

# ark_restore totals for overall success/failure tracking
ark_restore_attempts - counter
ark_restore_completions - counter
ark_restore_failures - counter

# ark_restore_seconds describes the duration of a backup from start to finish
ark_restore_seconds -- histogram
ark_restore_seconds_count -- counter
ark_backup_seconds_sum -- counter

# ark_restore_size describes the size of the resulting tarball
ark_restore_tarball_size_bytes - histogram
ark_restore_tarball_size_count - counter
ark_restore_tarball_size_sum - counter


# ark_restore_tarball_download_seconds describes the time to download the tarball
ark_restore_tarball_download_seconds
ark_restore_tarball_download_seconds_count
ark_restore_tarball_download_seconds_sum

# ark_restore_actions_seconds describes the time to run custom restorers
# Extra labels: restorer_name
ark_restore_restorers_seconds -- histograms
ark_restore_restorers_seconds_count -- counter
ark_restore_restorers_seconds_sum -- counter

```

* I did not include log file as I wasn’t sure how important that is @ncdc
* I did not include listing pods, services, etc because I don’t know if it matters the type of object, or just how long it takes to _get_ all the objects, regardless of type
* Do we want/need to track the duration of time it takes for a cloud-provider snapshots be taken?
* I’m not sure about the ark_restore_restorers — feedback welcome

**Comment by tomplus** [comment]
Thumbs up for the RED metrics.

**Comment by xamox** [comment]
RED metrics are something I'm looking for as well, I want to know if backups are failing most importantly, but all the other things mentioned would be a plus.

**Comment by corlettb** [comment]
+1

**Comment by ashish-amarnath** [comment]
Hoping that PR 531 will channel the discussion here and get us the metrics that we want and add more metrics in the future. Keeping the approach simple and as non-intrusive as possible. Feel free to leave feedback on the PR. I will continue to add the RED metrics that seem to be most useful for cluster operators.

**Comment by ashish-amarnath** [comment]
@ncdc You can assign this issue to me as I am working on the PR anyway. I am unable to do it myself.

**Comment by ncdc** [comment]
Assigned!

**Comment by rosskukulinski** [comment]
Added another metric to my list: 

# ark_info used to provide version and other meta data about this Ark instance
# * {version:"v0.9.0"}
# * {branch:"master"}
ark_info - gauge # always set to 1

**Comment by RochesterinNYC** [comment]
Tried out the `v0.9.0-alpha.2` Ark with the Prom metrics endpoint support added in https://github.com/heptio/ark/pull/531 (getting the metrics at `:8085/metrics` works like a charm).

However, with the above metric list/metrics in place now, as the `ark_backup_failure_total` counter metric is a total, we're not sure how to actually do alerting on this type of metric/the available metrics for failed backups. After one failed backup, this `ark_backup_failure_total` counter would always be `>=1` so alert conditions like "Alert me when `ark_backup_failure_total` > 0" would always be active/trigger.

**Comment by jrnt30** [comment]
You could use the `delta` or `rate` as a way of evaluating how that number
has changed over time.

**Comment by RochesterinNYC** [comment]
@jrnt30 are you referring to the `rate` and `delta` functions in the context of Prometheus? We're unfortunately not using Prometheus. We have an in-house metrics system and we're writing integration scripts/daemons that poll Prometheus metrics endpoints for various components in our clusters and sends/writes those metrics to that in-house metrics system.

It sounds like with the Prometheus metrics paradigm, we wouldn't ever expect the Ark `/metrics` endpoint to return any kind of rate information and the layer that is processing/polling the Prometheus metrics would need to handle that?

**Comment by jrnt30** [comment]
That is true, yes.  Typically with counters in Prometheus they only go up
and the query is then responsible for analyzing the "impact" of that.
`rate` and `delta` are two ways of assessing "has this thing changed in my
time window" or "what is the rate at which this has changed".

Something similar would be able to be done with your in-house tooling, but
can I understand how, without this context, that would be a bit confusing.

Does that clarify things at all?

**Comment by RochesterinNYC** [comment]
Yup, that makes sense. With our lack of Prometheus and unwillingness to add all this rate/delta calculation logic into the metrics middle layer (between Prom endpoints and our in-house system), we're probably going to try to emit metrics for "time-since-last-backup" by looking at the GCS bucket backing the Ark backups.

**Comment by jrnt30** [comment]
That is one way of doing it for sure.  Another approach which I tool while
this was being ironed out was writing a small utility that used the Backup
CRDs in the cluster itself to determine what the state should be
considered.

A word of caution that we ran into is that if you don't have an indicator
of what it was the last time you checked then you won't have a good handle
on what the new number means potentially. In the end I have a feeling you
may end up putting in a similar amount of effort to checkpoint that as you
would simply using the prom metrics.  A hacky way of doing that would be to
just use a configmap to actually write your last known state back into the
cluster each time your process "scrapes" the default metrics.

**Comment by ashish-amarnath** [comment]
So the ideal way to set this up is to do aggregation, think `rate` in prometheus land, and use a look back window. Here is an example of how to setup your alerts
```
- alert: FailedBackup
  expr: rate(ark_backup_failure_total[62m]) > 0
  for: 30s
  labels:
    severity: "warning"

```

**Comment by munai-das** [comment]
@ashish-amarnath Hi let me know if you need help in this implementation. We too need this feature.

**Comment by ncdc** [comment]
@munai-das thanks! Please feel free to open a PR to add any metrics you'd like. I'd recommend starting small - pick a single metric or a group of related metrics and do a PR for that.

**Comment by tomplus** [comment]
I've just installed 0.9.0 and and it works very well. I use Prometheus and these metrics are enough for me to create alerts or diagnose problems. Thank you. :+1:

**Comment by vitobotta** [comment]
Can anyone please clarify how to use these metrics? I have installed Prometheus Operator (which includes Grafana) and then Velero from Helm but I don't know how to access/use the metrics. Thanks

**Comment by StianOvrevage** [comment]
@vitobotta It's because Velero adds some annotations to inform Prometheus about it's metrics. However, Prometheus now discourages the use of annotations in favor of ServiceMonitors and has disabled the scraping of the annotated targets by default.

But you can have a look at https://github.com/coreos/kube-prometheus/pull/16 on how to add the required config. I'm doing that now and I'll try to remember and give some feedback if it works for me or not :)

**Comment by archmangler** [comment]
Is there any complete documentation on monitoring velero (preferably via prometheus) ?
